haskell-bindings

e1528532 opened the pull request:
# Purpose
For the type system i am going to develop i also need to code a tool which checks types in specifications. As this is going to be developed in Haskell, this PR is a WIP of the forthcoming haskell bindings.

# Checklist

- [X] commit messages are fine (with references to issues)
- [X] I ran all tests and everything went fine
- [X] I added unit tests
- [X] affected documentation is fixed
- [X] I added code comments, logging, and assertions
- [X] meta data is updated (e.g. README.md of plugins)

@markus2330 this is currently a very low-level approach. It is basically a direct mapping so i have to encapsulate all the pointer work in the IO monad. This is necessary because our methods are not referentially transparent but depend on the actual key/keyset pointer.

I used cabal as the build tool (its part of the standard haskell platform) and c2hs to generate most of the binding code automatically (this tool parses the chs files i have specified and generates the actual binding code for it and also takes care about proper marshalling between haskell and c types, therefore i decided to choose it over doing everything from hand. I think that saved me quite some work).

Still some work to do however, not everything is tested yet and the plugin-api is not ported (If i am even going to need it too)

markus2330 commented on src/bindings/haskell/libelektra-haskell.cabal.in:
```diff
@@ -0,0 +1,40 @@
+name:                libelektra-haskell
```
Is this file a duplicate of  src/bindings/haskell/libelektra-haskell.cabal?
Is  src/bindings/haskell/libelektra-haskell.cabal really needed?

e1528532 replied:
no it is not needed, it's just the cabal file used for my local development purposes, excluded it.

markus2330 commented on src/bindings/haskell/src/Elektra/KDB.chs:
```diff
@@ -0,0 +1,19 @@
+-- KDB METHODS
+-- ***
+
+-- TODO do we need this?
```
No.

markus2330 commented on src/bindings/haskell/test/Elektra.hs:
```diff
@@ -0,0 +1,63 @@
+      kdb <- kdbOpen parent
+      keyNew haskellPersisted >>= ksAppendKey ks
+      kdbGet kdb ks parent
+      kdbSet kdb ks parent
```
Does this only write to `/tests/<testname>`? Please also tag the KDB test with kdbtests, see doc/TESTING.md.

markus2330 commented:
@picca Are you still interested in a haskell binding? Do you want to do a code review?

> plugin-api is not ported (If i am even going to need it too)

Do you mean a possibility that plugins can be written in Haskell? That would be very nice. (If the Haskell run-time supports proper shutdown and so on...)

e1528532 commented on src/bindings/haskell/README.md:
```diff
@@ -0,0 +1,16 @@
+## Usage
+The bindings are built with cabal, Haskell's default build system. It needs a recent GHC version (>8),
+cabal and C2HS, which generates the actual bindings out of a metacode and will be called during the build
+automatically (Call `cabal install c2hs` to install it globally).
```
i haven't tried it out with earlier versions, it might work as well, will check before merging

e1528532 commented:
I think the tests demonstrate quite well how one can use the bindings, how would such example look like?

I have finally managed to add plugin support for haskell, and created a "template" plugin called haskelltemplate which demonstrates how the whole procedure works and acts as a blueprint for actual plugins, because there is some boilerplate code involved to deal with the boundaries between c and haskell. The most difficult part was getting the linking correctly.

Our add_plugin macro compiles via the c compiler whereas usually one would compile c libraries which include haskell code via the ghc then. This way ghc would have already set all the haskell-related linking parameters by itself, though it would have probably needed some elektra-related configuration then. I decided to stick to our way and figured out all the required libraries after some time. I've noticed the resulting plugin library is 12MB in size though for my simple example project, i suppose by default the whole haskell runtime has been linked statically to the plugin. At least that would have the advantage that it could run on systems without a haskell installation ;)
Also I have to rethink the dependency management (therefore the current build failure), right now it depends on having the bindings already installed and a few other hacks, though i tried to keep it as general as possible and avoided to hardcode paths.

I propose to finish this first version up so it can be merged soon, though i'd keep it experimental for now. I intend to keep it simple first and extend it later on in case i miss things for my actual project.

markus2330 commented:
> I think the tests demonstrate quite well how one can use the bindings, how would such example look like?

See for example `src/bindings/swig/python/examples`. Even better than example would be a tutorial how to use it.

> Also I have to rethink the dependency management (therefore the current build failure), right now it depends on having the bindings already installed and a few other hacks, though i tried to keep it as general as possible and avoided to hardcode paths.

Yes, I am all for a merge to happen soon but as minimal goal (even for experimental) the build should not never fail (if something is missing, disable the binding/plugin).

markus2330 commented:
Great to see the haskell plugin working! If you can find some way to avoid a template copied around, but instead only require haskell code to be written to implement a haskell plugin, it would be much better for maintenance. Maybe it is somehow possible to let haskell plugin developers to add something like `add_haskell_plugin (plugin_name haskell_sources...)` within the CMakeLists.txt to register a new haskell plugin? Then a haskell devs could also share code between plugins and you could put boiler plate code separately.

Then the plugin could be renamed to `src/plugins/haskell`.

The 12MB size is very low priority. First we should have some useful plugins before we optimize anything ;)

e1528532 commented:
Ok, I'd just make a haskell version of the python examples then. Of course we shouldn't merge it as long as it fails, thats what i meant with finishing it up.

I like your idea of having the haskell plugin as a base without having to copy it around, in the latest commit you see a simple version of that. It uses this technique to compile itself too, the cmake files for haskell plugins get quite simple now (as can be seen for the cmake file of the haskell plugin itself).
Furthermore i tried creating another simple plugin locally which was perfectly possible then without c code by using the new cmake macro. Ok, the test code was still in c, but the wrapper that starts the haskell runtime gets woven in automatically.
There is still a little boilerplate code involved as you need to export the functions expected by the c wrapper in haskell, but at least you don't need to touch c yourself anymore.

e1528532 commented:
ported the python examples to haskell, except for the index-based stuff as working with indexes doesn't really fit the haskell philosophy (although possible). It was a good idea to do so, i have found and fixed a potential segfault, Furthermore i added some useful helpher methods and missing things i encountered on the way. It should give a good impression on how to use this binding in its current early stage.

I still have to do some polishing and improve the cmake build though

e1528532 commented:
I think this should be good for now and implements the basic binding & plugin functionality. Furthermore it should be integrated in the cmake build process and the tests seem to work fine too.

Do you have any further comments or does someone want to review it? I think a review should be fine enough in the next iteration of this plugin before it leaves the experimental state and is not necessary now. Most likely i will add further improvements once i actively use the bindings for my type checker. In case you don't have any comments left, i think it can be merged.

markus2330 commented:
Thank you, great job!

> In case you don't have any comments left, i think it can be merged.

I overlooked this, in future please also update the "ready to merge".
